Add PagerDuty output plugin

ranjib opened the pull request:
This PR add PagerDuty output plugin. Using this users can configure telegraf to send PagerDuty alert when a particular metric cross given threshold. Following is an example:
```toml
[[outputs.pagerduty]]
metric = "cpu"
field = "time_iowait"
expression = " > 1050"
``` 
Will forward the metric  (and create a PagerDuty alert) when cpu.time_iowait reached above 1050.
I'll update the changelog & main readme, after the initial review.
### Required for all PRs:

- [x] CHANGELOG.md updated
- [x] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)
- [x] README.md updated (if adding a new plugin)

sparrc commented on plugins/outputs/pagerduty/pagerduty.go:
```diff
@@ -0,0 +1,114 @@
+
+import (
+	"fmt"
+	pd "github.com/PagerDuty/go-pagerduty"
```
we might not need to bring in this dependency. Since all we're using is "CreateEvent", which is a fairly simple http POST request.

I would recommend simple copying Pager Duty's code for that api call and keeping it in-repo: https://github.com/PagerDuty/go-pagerduty/blob/master/event.go

(you can put a comment with a link to the license at the top of the file)

sparrc commented on plugins/outputs/pagerduty/pagerduty.go:
```diff
@@ -0,0 +1,114 @@
+}
+
+var sampleConfig = `
+## PagerDuty service key
```
these should be indented 2 spaces

sparrc commented on plugins/outputs/pagerduty/pagerduty.go:
```diff
@@ -0,0 +1,114 @@
+	Metric     string            `toml:"metric"`
+	Field      string            `toml:"field"`
+	Expression string            `toml:"expression"`
+	Tags       map[string]string `toml:"tags"`
```
Calling this map just "tags" won't work because telegraf will interpret this map as tags that should be added to the metrics. Maybe call it "TagFilter" and provide an example of how to use it in SampleConfig.

sparrc commented on plugins/outputs/pagerduty/pagerduty.go:
```diff
@@ -0,0 +1,114 @@
+		if err != nil {
+			return err
+		}
+		log.Println("Created PagerDuty event for metric: ", metric.Name())
```
dont log this

sparrc commented on plugins/outputs/pagerduty/pagerduty.go:
```diff
@@ -0,0 +1,114 @@
+	}
+	for _, metric := range metrics {
+		if !p.Match(metric) {
+			log.Println("Metric is not matched by threshold, skipping")
```
dont log this

sparrc commented on plugins/outputs/pagerduty/pagerduty.go:
```diff
@@ -0,0 +1,114 @@
+		m["name"] = metric.Name()
+		m["timestamp"] = metric.UnixNano() / 1000000000
+		event.Details = m
+		_, err := pd.CreateEvent(event)
```
Here you can get the incident_key back from the response (see https://v2.developer.pagerduty.com/docs/events-api#incident-de-duplication-and-incident_key and https://v2.developer.pagerduty.com/v2/docs/trigger-events)

Seems like once you make the first CreateEvent request, you should save the incident key so that we're not constantly creating new incidents for the same alert. I believe that a single alert should always be associated with the same incident?

ranjib commented:
@sparrc thanks a tonne for the review. I have addressed most of your comments. Including 0 external deps, and reusing existing incident key to trigger or resolve events.
Couple of things are still missing/good to have:
- More unit test coverage (I'll do this if you think the implementation is alright in its current form)
- With current implementation users can only configure one alert, or alert based on one metric, I want that to be changed to many. I know how to do it in the golang/plugin stuff, but was confused with the TOML -> config bits. I'll appreciate some help around it, i.e. any examples of TOML with nested config items in the plugin, what will be the TOML representation,
```go
type Alert struct {
  Metric string
  Expression string
  Field string
  TagFIlter map[string]string
}
type PD struct {
 ServiceKey
Alerts []alerts `toml:"alerts"`
}
```

ranjib commented on plugins/outputs/pagerduty/pagerduty.go:
```diff
@@ -0,0 +1,129 @@
+  field = "time_iowait"
+  ## Tag filter, when present only metrics with the specified tag value
+  ## will be considered for further processing
+  [[outputs.pagerduty.tag_filter]]
```
I am not sure if I am doing this right, I'll cross check it during weekend

francois2metz replied:
`tag_filter` should be a struct with 2 fields, `role` and `region`.

ranjib replied:
I have extracted this as Tag type with Name and Value field

francois2metz commented:
You can see other implementations of config with array in the snmp plugin: https://github.com/influxdata/telegraf/blob/master/plugins/inputs/snmp/snmp.go

burdandrei commented:
What is the point? Why to have alerts both is telegraf and kapacitor?

allen13 commented:
@burdandrei I run kapacitor in production and think it's a great product. However, it is yet another item to maintain in addition to being another point of failure. Giving all of your Telegraf instances the ability to send basic metrics alerts to PagerDuty directly is light weight and completely fault-tolerant.

burdandrei commented:
@allen13, right, is your telegraf instance didn't died =)

ranjib commented:
@francois2metz thanks for the review. I'll take a look and get those bits sorted.
@allen13 yes you can. But by design telegraf allows multiple input and output, and is not prescriptive about what an user should choose. I use mostly DataDog to PagerDuty alerting, and i think that is more idiomatic for 80% of use cases, but then there are cases where direct PD integration is desired. Specially where we dont have any secondary metric component or metric sevice -> PD integration is absent, or we want to reduce the latency between metric capture/shipping/PD alert trigger. Hence this PR.
I would close this if the authors feel that this is not in sync with telegraf's philosophy/functionality.

sparrc commented:
for me it's OK if we have some simple alerts in Telegraf, though I wonder if it warrants another type of plugin (`alert` plugins?) that would be optimized for streaming and analyzing data as it flows through, rather than (temporarily) storing and writing data.

ranjib commented:
@sparrc  that would be awesome. I can use that to store incident_key. If we mimic Nagios's alerting workflow, we can use the storage to detect flapping alerts. We can also use the storage to temporarily buffer alerts if the upstream endpoint is not available.

burdandrei commented:
@sparrc alert plugins is great idea!

sparrc commented:
@ranjib what is the status of this? do you want it to be merged still? do you want to wait for another type of `alert` plugins? If so, can you explain the use-case and what exactly you're currently not able to do with output plugins?

ranjib commented:
@sparrc I would love to get this merged. I have rebased it with master and fixed the filter_tag type. Earlier it was not working as expected (with map[string]string type).

FYI I am using this plugin to raise PD alert when serf membership changes (serf emits statsd metrics against telegraf statsd endpoint)
Let me know if theres any outstanding comments

ranjib commented:
@sparrc Not immediately, but in future having alert specific plugin might help, given there exist multiple providers. I am not sure what would be the best/common format. Happy to help any way I can.

ranjib commented:
@sparrc  bump

sparrc commented:
Hi @ranjib, first off wanted to thank you for this code contribution, but I'm afraid I simply can't merge this at the moment. I fear that turning Telegraf into an alerting tool, in addition to all the things it currently is, is going to open a whole new can of worms.

So for that reason, I'm going to have to reject the PR on the grounds that it will expand the scope of Telegraf into an unmaintainable state.

I hope we can get some simple alerting capability into telegraf at a later date.
